.github/workflows/build.yml

```diff
@@ -0,0 +1,227 @@
+          pkg_name=${{ env.PKG_NAME }}
+          pkg_metadata=${{ env.METADATA }}
+          echo "pkg-metadata=${pkg_metadata}" | tee -a "${GITHUB_OUTPUT}"
+          echo "pkg-name=${pkg_name}" | tee -a "${GITHUB_OUTPUT}"
+      - name: Decide version number
+        id: get-product-version
+        shell: bash
+        run: echo "product-version=$(cat version/VERSION)" >> $GITHUB_OUTPUT
+      - name: Report chosen version number
+        run: |
+          [ -n "${{steps.get-product-version.outputs.product-version}}" ]
+          echo "::notice title=Waypoint Version::${{ steps.get-product-version.outputs.product-version }}"
+
+  get-go-version:
+    name: "Determine Go toolchain version"
+    runs-on: ubuntu-latest
+    outputs:
+      go-version: ${{ steps.get-go-version.outputs.go-version }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Determine Go version
+        id: get-go-version
+        run: |
+          echo "Building with Go $(cat .go-version)"
+          echo "go-version=$(cat .go-version)" >> $GITHUB_OUTPUT
+
+  generate-metadata-file:
+    name: "Generate release metadata"
+    runs-on: ubuntu-latest
+    needs: get-product-version
+    outputs:
+      filepath: ${{ steps.generate-metadata-file.outputs.filepath }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Generate package metadata
+        id: generate-metadata-file
+        uses: hashicorp/actions-generate-metadata@v1
+        with:
+          version: ${{ needs.get-product-version.outputs.product-version }}
+          product: ${{ env.PKG_NAME }}
+      - uses: actions/upload-artifact@v2
+        with:
+          name: metadata.json
+          path: ${{ steps.generate-metadata-file.outputs.filepath }}
+
+  build-ui:
+    runs-on: ubuntu-latest
+    name: Build UI
+    needs: get-product-version
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up node and yarn
+        uses: actions/setup-node@v3
+        with:
+          node-version: 14
+          cache: yarn
+          cache-dependency-path: ui/yarn.lock
+      - name: Build UI
+        run: |
+          export PKG_VERSION=${{ needs.get-product-version.outputs.product-version }}
+          export PKG_BINARY_TYPE=${{ needs.get-product-version.outputs.pkg-metadata }}
+          export PKG_COPYRIGHT_YEAR=$(git show -s --format=%cd --date=format:%Y HEAD)
+          export PKG_NAME=${{ needs.get-product-version.outputs.pkg-name }}
+          echo "${PKG_NAME} version is ${PKG_VERSION}"
+          echo "${PKG_NAME} binary type is ${PKG_BINARY_TYPE}"
+          echo "${PKG_NAME} copyright year is ${PKG_COPYRIGHT_YEAR}"
+          cd ui && make && cd ..
+      - name: Save assets to cache
+        uses: actions/cache/save@v3
+        with:
+          path: ui/dist/assets
+          key: ui_assets_${{ github.sha }}
+
+  build-assets:
+    # FIXME: can releases be tagged as per CRT standards (ie at the end of the pipeline)?
+    runs-on: ubuntu-latest
+    name: Build Assets
```

Comment:
The current `waypoint-releases` repo seems to tag the commit at the beginning of the pipeline as seen [here](https://github.com/hashicorp/waypoint-releases/blob/main/.circleci/config.yml#L368). However, CRT typically tags releases on `promote-to-production` - is @waypoint ok with RelEng tagging releases at the end of pipeline?